### PR TITLE
harfbuzz: cross compilation

### PR DESCRIPTION
**Pull request by Mindavi**
###### Motivation for this change

Support cross-compilation by disabling introspection. Tested by building for aarch64-multiplatform.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/libraries/harfbuzz/default.nix [hunk 1]
```diff
@@ -43,24 +57,28 @@ stdenv.mkDerivation {
...
+    glib # glib-mkenums
+  ] ++ lib.optional (withIntrospection) gobject-introspection;
 
   buildInputs = [ glib freetype cairo ] # recommended by upstream
+    ++ lib.optional (withIntrospection) gobject-introspection
     ++ lib.optionals withCoreText [ ApplicationServices CoreText ];
 
-  propagatedBuildInputs = []
+  propagatedBuildInputs = [ ]
     ++ optional withGraphite2 graphite2
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/harfbuzz/default.nix:82]:
```suggestion
  propagatedBuildInputs = optional withGraphite2 graphite2
```

**Comment by sternenseemann** [comment]
Is it impossible to cross-compile `gobject-introspection` or with `gobject-introspection`-support? Or is it just not possible *in nixpkgs*. A comment explaining the situation would be helpful.

**Comment by Mindavi** [comment]
In the future it might become possible due to the work of @Ericson2314 here: https://github.com/NixOS/nixpkgs/pull/88222. See also the related issue: https://github.com/NixOS/nixpkgs/issues/72868.

However, in the present day it's not possible and I think John Ericson is working hard to get the support upstream to prevent having to maintain a fork. It seems like upstream is not very willing to accept the patches though, so the process seems to be slow. As you can see in the PRs, the support is also not complete, so there's some work to do on that.

Any comments on this @Ericson2314 ? This is just my own interpretation of what's happening.

So in short: it's not possible right now, but might be in the future.

**Comment by sternenseemann** [comment]
Ah, thanks for the explanation! Could you link the related issue or PR with a short note in a comment? I think this makes it clearer to the reader what is going on.

**Comment by Mindavi** [comment]
I'd rather put it in the commit message to prevent adding that comment on all places that disable introspection (there are at least ~15 or so). But either way is fine by me :).

**Comment by sternenseemann** [comment]
Sure!

**Comment by sternenseemann** [comment]
@ofborg build pkgsCross.aarch64-multiplatform.harfbuzz

File: pkgs/development/libraries/harfbuzz/default.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/development/libraries/harfbuzz/default.nix:81]: I actually kinda like this style, easier to rearrange lines later.

File: pkgs/development/libraries/harfbuzz/default.nix [hunk 2]
```diff
@@ -1,13 +1,26 @@
...
-, icu, graphite2, harfbuzz # The icu variant uses and propagates the non-icu one.
-, ApplicationServices, CoreText
+, icu
+, graphite2
+, harfbuzz # The icu variant uses and propagates the non-icu one.
+, ApplicationServices
+, CoreText
 , withCoreText ? false
 , withIcu ? false # recommended by upstream as default, but most don't needed and it's big
 , withGraphite2 ? true # it is small and major distros do include it
```
Comment (Ericson2314) [inline pkgs/development/libraries/harfbuzz/default.nix:19]:
```suggestion
, withGraphite2 ? true # it is small and major distros do include it
, withIntrospection ? stdenv.buildPlatform == stdenv.hostPlatform
```

File: pkgs/development/libraries/harfbuzz/default.nix [hunk 3]
```diff
@@ -43,8 +56,11 @@ stdenv.mkDerivation {
     (mesonFeatureFlag "graphite" withGraphite2)
     (mesonFeatureFlag "icu" withIcu)
     (mesonFeatureFlag "coretext" withCoreText)
+    (mesonFeatureFlag "introspection" (stdenv.buildPlatform == stdenv.hostPlatform))
```
Comment (Ericson2314) [inline pkgs/development/libraries/harfbuzz/default.nix:59]:
```suggestion
    (mesonFeatureFlag "introspection" withIntrospection)
```

File: pkgs/development/libraries/harfbuzz/default.nix [hunk 4]
```diff
@@ -58,9 +74,10 @@ stdenv.mkDerivation {
   ];
 
   buildInputs = [ glib freetype cairo ] # recommended by upstream
-    ++ lib.optionals withCoreText [ ApplicationServices CoreText ];
+    ++ lib.optionals withCoreText [ ApplicationServices CoreText ]
+    ++ lib.optionals (stdenv.buildPlatform == stdenv.hostPlatform) [ gobject-introspection ];
```
Comment (Ericson2314) [inline pkgs/development/libraries/harfbuzz/default.nix:78]:
```suggestion
    ++ lib.optionals withIntrospection [ gobject-introspection ];
```

**Comment by Ericson2314** [comment]
Yes, I am blocked for gobject-introspection on upstream merging PRs for like a year now: https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/227 and https://gitlab.gnome.org/GNOME/gobject-introspection/-/merge_requests/224. They even approved the first one. No idea what the deal is; it's been frustrating.

File: pkgs/development/libraries/harfbuzz/default.nix (hunk 4 above)
Comment (Mindavi) [reply pkgs/development/libraries/harfbuzz/default.nix:78]: It seems that you and @jtojnar disagree on the style here. What shall I do here? I personally find the `withIntrospection` nice since it's also easy to search for if the introspection stuff gets updated to support cross.

But I don't have a strong opinion on what's best.

Comment (jtojnar) [reply pkgs/development/libraries/harfbuzz/default.nix:78]: Yeah, I am trying to reduce the noise in the expressions I maintain – the chance that someone will want to disable introspection aside from cross-compiling is pretty small so the additional clutter is not worth it IMO. But maintainers of each package will have to choose.

File: pkgs/development/libraries/harfbuzz/default.nix (hunk 2 above)
Comment (jtojnar) [reply pkgs/development/libraries/harfbuzz/default.nix:19]: Alternately,

you can define `withIntrospection` in a `let` block as not to clutter the inputs with stuff noone will override anyway.

File: pkgs/development/libraries/harfbuzz/default.nix (hunk 4 above)
Comment (Ericson2314) [reply pkgs/development/libraries/harfbuzz/default.nix:78]: @jtojnar says let bind it rather than make it a parameter. I don't care strongly about that; I just want a variable to be somehow bound (let or parameter) with a descriptive name, rather than repeating the cross check which isn't inherently related to introspection.

File: pkgs/development/libraries/harfbuzz/default.nix (hunk 1 above)
Comment (Mindavi) [reply pkgs/development/libraries/harfbuzz/default.nix]: Done.

File: pkgs/development/libraries/harfbuzz/default.nix (hunk 4 above)
Comment (Mindavi) [reply pkgs/development/libraries/harfbuzz/default.nix]: Thanks for the tip! Updated with a let-expression.

**Comment by Mindavi** [comment]
Updated to address all comments. Glad we could find a nice solution. Thanks for reviewing! (and great to see that some stuff from Ericson2314 has finally been merged upstream on gobject-introspection, great step in the right direction :smile:).

File: pkgs/development/libraries/harfbuzz/default.nix [hunk 5]
```diff
@@ -1,29 +1,43 @@
...
+, docbook-xsl-nons
+, docbook_xml_dtd_43
 }:
 
 let
   version = "2.8.0";
   inherit (lib) optional optionals optionalString;
   mesonFeatureFlag = opt: b:
     "-D${opt}=${if b then "enabled" else "disabled"}";
+  withIntrospection = stdenv.buildPlatform == stdenv.hostPlatform;
```
[the remark this answers is not in the document]
Comment (Mindavi) [reply pkgs/development/libraries/harfbuzz/default.nix]: Fine with me.

**Review by sternenseemann — APPROVED** [review]
